target/ppc/translate/vsx-impl.c.inc

```diff
@@ -2124,44 +2124,60 @@ static void gen_xvxsigdp(DisasContext *ctx)
     tcg_temp_free_i64(xbl);
 }
 
-#define XVIGER(name, func_name, op_mask)                                \
-    static bool trans_##name (DisasContext *ctx, arg_##name * a)        \
+/*
+ * 0x8 SATURATION BIT
+ * 0x4 ACCUMULATE BIT
+ * 0x2 POSITIVE MULTIPLY BIT
+ * 0x1 POSTIIVE ACCUMULATE BIT
+ */
+#define XVIGER(name, func_name, saturate, accumulate)                   \
+    static bool trans_##name (DisasContext *ctx, arg_XX3 *a)            \
     {                                                                   \
-        uint64_t mask;                                                  \
+        uint32_t mask, op_flag;                                         \
         REQUIRE_INSNS_FLAGS2(ctx, ISA310);                              \
+        if (unlikely((a->xa >= a->at*4 && a->xa <= a->at*4 + 3) ||      \
+                     (a->xb >= a->at*4 && a->xb <= a->at*4 + 3))) {     \
+            return false;                                               \
+        }                                                               \
                                                                         \
-        mask = (uint64_t)0xFFFFFFFF;                                    \
-        mask |= ((uint64_t)op_mask) << 32;                              \
-        gen_helper_##func_name (cpu_env, tcg_constant_i32(a->ax),       \
-                                         tcg_constant_i32(a->bx),       \
-                                         tcg_constant_i32(a->tx),       \
-                                         tcg_constant_i64(mask));       \
+        mask = (uint32_t)0xFFFFFFFF;                                    \
+        op_flag = ((saturate << 1) | accumulate) << 2;                  \
```

Comment:
Having nested shifts it is harder to know which bits are being set (``saturate`` is being shifted 1 bit, and then 2 more bits).
IMO, it would be simpler to read:
```
op_flag = (saturate << 3) | (accumulate << 2);
```